tests/test_table.py

```diff
@@ -1204,6 +1211,153 @@ def test_map(table, map):
     pd.testing.assert_frame_equal(result, expected)
 
 
+@pytest.mark.parametrize("storage_format", ["csv", "parquet"])
+def test_hash(tmpdir, storage_format):
```

Comment:
I wonder whether the tests of hashing would not be better refactored into a test class, with one dedicated test for each assertion?

I am asking because it is hard to read and understand whether the assertions are based on chained transformations of the database or whether they each rely on the saved database.

Reply:
I'm not that familiar with using test classes in `pytest`. Is the following what you would have in mind?

```python
@pytest.mark.parametrize("storage_format", ["csv", "parquet"])
class TestHash:
    r"""Test if PARQUET file hash changes with table.

    We store a MD5 sum associated with the dataframe,
    that was used to create the file,
    in the metadata of the PARQUET file.
    Those MD5 sum is supposed to change,
    if any of the table rows, (index) columns changes,
    the data type of the entries changes,
    or the name of a column changes.

    Args:
        tmpdir: tmpdir fixture
        storage_format: storage format of table file

    """

    def db(self, tmpdir, storage_format):
        r"""Create minimal database with scheme and table."""
        self.db_root = audeer.path(tmpdir, "db")
        self.storage_format = storage_format
        self.table_file = audeer.path(self.db_root, f"db.table.{storage_format}")
        db = audformat.Database("mydb")
        db.schemes["int"] = audformat.Scheme("int")
        index = audformat.segmented_index(["f1", "f2"], [0, 1], [1, 2])
        db["table"] = audformat.Table(index)
        db["table"]["column"] = audformat.Column(scheme_id="int")
        db["table"]["column"].set([0, 1])
        db.save(self.db_root, storage_format=self.storage_format)
        return db

    def md5(self) -> str:
        r"""Get MD5 sum for table file."""
        ext = self.storage_format
        if ext == "csv":
            md5 = audeer.md5(self.table_file)
        elif ext == "parquet":
            md5 = parquet.read_schema(self.table_file).metadata[b"hash"].decode()
        return md5

    def test_change_index(self, tmpdir, storage_format):
        r"""Change table index."""
        db = self.db(tmpdir, storage_format)
        md5 = self.md5()
        index = audformat.segmented_index(["f1", "f1"], [0, 1], [1, 2])
        db["table"] = audformat.Table(index)
        db["table"]["column"] = audformat.Column(scheme_id="int")
        db["table"]["column"].set([0, 1])
        db.save(self.db_root, storage_format=self.storage_format)
        assert self.md5() != md5

    def test_change_column_name(self, tmpdir, storage_format):
        r"""Change table column name."""
        db = self.db(tmpdir, storage_format)
        md5 = self.md5()
        index = audformat.segmented_index(["f1", "f2"], [0, 1], [1, 2])
        db["table"] = audformat.Table(index)
        db["table"]["col"] = audformat.Column(scheme_id="int")
        db["table"]["col"].set([0, 1])
        db.save(self.db_root, storage_format=self.storage_format)
        assert self.md5() != md5

    def test_change_column_order(self, tmpdir, storage_format):
        r"""Change order of table columns."""
        db = self.db(tmpdir, storage_format)
        index = audformat.segmented_index(["f1", "f2"], [0, 1], [1, 2])
        db["table"] = audformat.Table(index)
        db["table"]["col1"] = audformat.Column(scheme_id="int")
        db["table"]["col1"].set([0, 1])
        db["table"]["col2"] = audformat.Column(scheme_id="int")
        db["table"]["col2"].set([0, 1])
        db.save(self.db_root, storage_format=self.storage_format)
        md5 = self.md5()
        db["table"] = audformat.Table(index)
        db["table"]["col2"] = audformat.Column(scheme_id="int")
        db["table"]["col2"].set([0, 1])
        db["table"]["col1"] = audformat.Column(scheme_id="int")
        db["table"]["col1"].set([0, 1])
        db.save(self.db_root, storage_format=self.storage_format)
        assert self.md5() != md5

    def test_change_row_order(self, tmpdir, storage_format):
        r"""Change order of table rows."""
        db = self.db(tmpdir, storage_format)
        md5 = self.md5()
        index = audformat.segmented_index(["f2", "f1"], [1, 0], [2, 1])
        db["table"] = audformat.Table(index)
        db["table"]["column"] = audformat.Column(scheme_id="int")
        db["table"]["column"].set([1, 0])
        db.save(self.db_root, storage_format=storage_format)
        assert self.md5() != md5

    def test_change_values(self, tmpdir, storage_format):
        r"""Change table values."""
        db = self.db(tmpdir, storage_format)
        md5 = self.md5()
        index = audformat.segmented_index(["f1", "f2"], [0, 1], [1, 2])
        db["table"] = audformat.Table(index)
        db["table"]["column"] = audformat.Column(scheme_id="int")
        db["table"]["column"].set([1, 0])
        db.save(self.db_root, storage_format=self.storage_format)
        assert self.md5() != md5

    def test_copy_table(self, tmpdir, storage_format):
        r"""Replace table with identical copy."""
        db = self.db(tmpdir, storage_format)
        md5 = self.md5()
        table = db["table"].copy()
        db["table"] = table
        db.save(self.db_root, storage_format=self.storage_format)
        assert self.md5() == md5
```

Reply:
Yes,  test classes in pytest afaik do not do much more than improving readability through isolation. 
Now it becomes clear that each time `db.save`  is called a different tmpdir is setup and that the transactions do not depend on each other - so at least for me - it is much nicer to read: we can quickly see that it is robust or how it behaves against copying, column exchange and so on. 

Once you commit and run the tests the conversation can be resolved - I am anticipating by pressing the "reolve conversation" button right now.

Reply:
OK, great, I have pushed my changes.